### ISSUE TITLE
Support lazy loading

### ISSUE THREAD

**Issue by RubenVerborgh** [LinkedSoftwareDependencies/Components.js#10]
It could be a great speed-up if software components are only loaded when needed.

Fully implementing this might require architectural changes, but perhaps some tweaks like [this one](https://github.com/RubenVerborgh/N3.js/blob/v0.11.2/N3.js#L16) could be implemented.

**Comment by rubensworks** [comment]
Related to dynamic actor loading in comunica: https://github.com/comunica/comunica/issues/9

**Comment by RubenVerborgh** [comment]
So I've been thinking that this is really easy for constructors that take the _hash_ approach, i.e., `new MyClass(options)`. In that case, we can make `options` a lazy object, in which properties are initialized on first access.

However, a problem might be then if `MyClass` simply does:
```
constructor (options) {
  this._a = options.a;
  this._b = options.b;
}
```
which would then instantiate all of them.

Maybe we should consider `this._options = options` instead, and access only when needed.

**Comment by rubensworks** [comment]
In the case of Comunica, the first approach is however taken. See https://github.com/comunica/comunica/blob/master/packages/core/lib/Actor.ts#L37
We could however make an abstraction that results in the same, but uses the `args` as source of options in the background. Will play around with this a bit soon.

**Comment by RubenVerborgh** [comment]
> We could however make an abstraction that results in the same, but uses the args as source of options in the background. Will play around with this a bit soon.

That's pretty cool indeed. Kind of a “soft assign” then.

**Comment by rubensworks** [comment]
For reference, the async `import()` may also be useful in some cases.

For example, certain comunica actors only require certain dependencies in their run-phase (like the heavy-weight JSON-LD lib), which means that those dependencies could be `import()`-ed asynchronously in their run function.

**Comment by RubenVerborgh** [comment]
Nice. One of my dreams is that, in this example, the JSON-LD lib is only downloaded when needed.

**Comment by rubensworks** [comment]
Actually, we can implement support for lazy loading pretty easily _and_ maintain backwards compatibility by allowing a `lazy` field to be set to `true` during parameter specification.

For example, a component could have the following parameter:
```
{ "@id": "ex:MyModule/MyComponent#thing", "unique": true, "lazy": true }
```

When this component is instantiated, the value for `ex:MyModule/MyComponent#thing` will not be a `Thing` value directly, but a `() => Promise<Thing>`.

Behind the scenes, Components.js could use `import()` when components are passed instead of raw values.

> One of my dreams is that, in this example, the JSON-LD lib is only downloaded when needed.

_Should_ be possible/configurable with Webpack once this works in a Node environment :-)

**Comment by RubenVerborgh** [comment]
> allowing a lazy field to be set to true during parameter specification.

That's neat!

**Comment by rubensworks** [comment]
Implemented first version in 411c5d2c6d86a76e8b06bf75046253d54a2d6704.
